File: app/src/main/java/ru/roscha_akademii/medialib/common/MediaLibApplication.java
package ru.roscha_akademii.medialib.common;

import android.app.Application;

import com.crashlytics.android.Crashlytics;
import com.squareup.leakcanary.LeakCanary;
import com.squareup.leakcanary.RefWatcher;

import io.fabric.sdk.android.Fabric;

public class MediaLibApplication extends Application {
    private RefWatcher refWatcher;

    @Override
    public void onCreate() {
        super.onCreate();
        refWatcher = LeakCanary.install(this);
        Fabric.with(this, new Crashlytics());
    }

    protected ApplicationComponent component;

    private ApplicationComponent createComponent() {
        return DaggerApplicationComponent
                .builder()
                .androidModule(new AndroidModule(this, refWatcher))
                .build();
    }

    public ApplicationComponent component() {
        if (component == null) {
            synchronized (this) {
                if (component == null) {
                    component = createComponent();
                }
            }
        }

        return component;
    }

    public RefWatcher refWatcher() {
        return refWatcher;
    }

}

File: app/src/main/java/ru/roscha_akademii/medialib/common/MediaLibApplication.kt
package ru.roscha_akademii.medialib.common

import android.os.StrictMode
import android.support.multidex.MultiDexApplication
import com.crashlytics.android.BuildConfig
import com.crashlytics.android.Crashlytics
import com.squareup.leakcanary.LeakCanary
import com.squareup.leakcanary.RefWatcher
import io.fabric.sdk.android.Fabric
import net.danlew.android.joda.JodaTimeAndroid

open class MediaLibApplication : MultiDexApplication() {
    protected var _component: ApplicationComponent? = null
    val component: ApplicationComponent
        get() {
            return _component!!
        }

    lateinit var refWatcher: RefWatcher

    override fun onCreate() {
        if (BuildConfig.DEBUG) {
            StrictMode.setThreadPolicy(
                    StrictMode.ThreadPolicy.Builder()
                            .detectAll()
                            .penaltyLog()
                            .penaltyDialog()
                            .penaltyDeathOnNetwork()
                            .build()
            )
            StrictMode.setVmPolicy(
                    StrictMode.VmPolicy.Builder()
                            .detectAll()
                            .penaltyLog()
                            .penaltyDeath()
                            .build()
            )
        }

        super.onCreate()

        JodaTimeAndroid.init(this)

        refWatcher = LeakCanary.install(this)

        if (_component == null) {
            _component = DaggerApplicationComponent
                    .builder()
                    .androidModule(AndroidModule(this, refWatcher))
                    .build()
        }

        Fabric.with(this, Crashlytics())
    }
}
